packages/snfoundry/contracts/src/challenge1.cairo

```diff
@@ -1,39 +1,141 @@
+            self._not_completed();
+            assert(get_block_timestamp() >= self.deadline.read(), 'Staking period has not ended');
+            let staked_amount = self.token.read().balanceOf(get_contract_address());
+            if (staked_amount >= THRESHOLD) {
+                self._complete_transfer(staked_amount);
+            } else {
+                self.open_for_withdraw.write(true);
+            }
+        }
+
+        fn withdraw(ref self: ContractState) {
+            // Todo in UI: Staker contract should approve to transfer back the sender_amount to the sender
+            assert(self.open_for_withdraw.read(), 'Withdrawal not allowed');
+            let sender = get_caller_address();
+            let sender_amount = self.balances.read(sender);
+            assert(sender_amount > 0, 'No balance to withdraw');
+            self.token.read().transferFrom(get_contract_address(), sender, sender_amount);
+            self.balances.write(sender, 0);
+        }
+
+        fn balances(self: @ContractState, account: ContractAddress) -> u256 {
+            self.balances.read(account)
+        }
+
+        fn total_balance(self: @ContractState) -> u256 {
+            self.balances.read(get_contract_address())
+        }
+
+        fn deadline(self: @ContractState) -> u64 {
+            self.deadline.read()
+        }
+
+        fn threshold(self: @ContractState) -> u256 {
+            THRESHOLD
+        }
+
+        fn open_for_withdraw(self: @ContractState) -> bool {
+            self.open_for_withdraw.read()
+        }
+
+        fn example_external_contract(self: @ContractState) -> ContractAddress {
+            self.external_contract_address.read()
+        }
+
+        fn completed(self: @ContractState) -> bool {
+            let external_contract = IExampleExternalContractDispatcher {
+                contract_address: self.external_contract_address.read()
+            };
+            external_contract.completed()
+        }
+
+        fn time_left(self: @ContractState) -> u64 {
+            let current_time = get_block_timestamp();
+            let deadline = self.deadline.read();
+            if current_time < deadline {
+                deadline - current_time
+            } else {
+                0
+            }
+        }
+    }
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn _complete_transfer(ref self: ContractState, amount: u256) {
+            let external_contract_address = self.external_contract_address.read();
+            let external_contract_dispatcher = IExampleExternalContractDispatcher {
+                contract_address: external_contract_address
+            };
+            // Todo in UI: Staker contract should approve to transfer the staked_amount to the external contract
+            // self
+            //     .token
+            //     .read()
+            //     .transferFrom(get_contract_address(), external_contract_address, amount);
```

Comment:
here you directly transfer the amount the contract holds, dont need to use `transferFrom()`, can use `transfer()`

Reply:
ok, but we still need to test it with the required approve()